semgrep-core/src/engine/Match_rules.ml

```diff
@@ -801,36 +785,57 @@ and range_matches_with_pattern env r mvar opt_lang formula =
               in
               nested_formula_has_matches env formula lazy_ast_and_errors
                 lazy_content
-                (Some { r with r = mval_range }) ) )
-
-and lazy_ast_of_string lang base_tok str warn_msg =
-  let fix_toks ast =
-    (* We put `str` into a temporary file and parse it as `lang`, then all
-     * location info refers to this temporary file. But we want parse errors
-     * and matching ranges to refer to positions in the original file! So, we
-     * apply Parse_info.adjust_info_wrt_base to every parse info to fix this. *)
-    let base_loc = Parse_info.token_location_of_info base_tok in
-    let visitor =
-      Map_AST.mk_visitor
-        {
-          Map_AST.default_visitor with
-          Map_AST.kinfo =
-            (fun (_, _) tok -> Parse_info.adjust_info_wrt_base base_loc tok);
-        }
-    in
-    visitor.Map_AST.vprogram ast
-  in
-  lazy
-    (let ext =
-       match Lang.ext_of_lang lang with x :: _ -> x | [] -> assert false
-     in
-     Common2.with_tmp_file ~str ~ext (fun file ->
-         let { Parse_target.ast; errors; _ } =
-           Parse_target.parse_and_resolve_name_use_pfff_or_treesitter lang file
-         in
-         let ast = fix_toks ast in
-         if errors <> [] then pr2 (warn_msg ());
-         (ast, errors)))
+                (Some { r with r = mval_range }) )
+      | Some xlang, MV.Text (content, _tok)
+      | Some xlang, MV.E (G.L (G.String (content, _tok))) ->
+          (* We reinterpret the matched text as `xlang`. *)
+          let ext =
```

Comment:
I don't think the extension really matters.  I don't think Parse_target will reject files with the wrong extension